src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,1425 @@
+    json_t *name;
+    size_t index;
+
+    inp = json_load_file("issuers", 0, &json_error);
+    if (inp == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to parse issuers file on line [%d]: [%s].\n",
+              json_error.line, json_error.text);
+        ret = EINVAL;
+        goto done;
+    }
+
+    issuers = json_object_get(inp, "issuers");
+    if (!json_is_array(issuers)) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Issuers is not an array.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = ENOENT;
+    json_array_foreach(issuers, index, i) {
+        name = json_object_get(i, "name");
+        if (!json_is_string(name)) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Malformed issuer name.\n");
+            continue;
+        }
+        if (strcmp(json_string_value(name), issuer_name) == 0) {
+            ret = get_endpoints(i, dc_ctx);
+            break;
+        }
+    }
+
+done:
+    json_decref(issuers);
+
+    return ret;
+}
+
+static errno_t set_endpoints(struct devicecode_ctx *dc_ctx,
+                             const char *device_auth_endpoint,
+                             const char *token_endpoint,
+                             const char *userinfo_endpoint,
+                             const char *jwks_uri,
+                             const char *scope)
+{
+    int ret;
+
+    dc_ctx->device_authorization_endpoint = talloc_strdup(dc_ctx,
+                                                          device_auth_endpoint);
+    if (dc_ctx->device_authorization_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing device_authorization_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->token_endpoint = talloc_strdup(dc_ctx, token_endpoint);
+    if (dc_ctx->token_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing token_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->userinfo_endpoint = talloc_strdup(dc_ctx,userinfo_endpoint);
+    if (dc_ctx->userinfo_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing userinfo_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    if (jwks_uri != NULL && *jwks_uri != '\0') {
+        dc_ctx->jwks_uri = talloc_strdup(dc_ctx, jwks_uri);
+        if (dc_ctx->jwks_uri == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Missing jwks_uri.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    if (scope != NULL && *scope != '\0') {
+        dc_ctx->scope = talloc_strdup(dc_ctx, scope);
+        if (dc_ctx->scope == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Missing scopes.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    ret = EOK;
+done:
+    return ret;
+}
+
+/* The curl write_callback will always append the received data. To start a
+ * new string call clean_http_data() before the curl request.*/
+static void clean_http_data(struct devicecode_ctx *dc_ctx)
+{
+    talloc_free(dc_ctx->http_data);
```

Comment:
also memset to 0 since it contains user pin?

Reply:
'memset to 0' won't do the trick to erase sensitive data properly.

`util.h` has `sss_erase_talloc_mem_securely` and `sss_erase_mem_securely` helpers (that use `explicit_bzero()` if available or a trick with `volatile` otherwise to avoid compiler optimization)

Reply:
Ah, this is already used in the latest version.